extension/src/plots/model/custom.ts

```diff
@@ -16,28 +18,93 @@ export const isCheckpointValue = (
 export const isCheckpointPlot = (plot: CustomPlot): plot is CheckpointPlot =>
   plot.type === CustomPlotType.CHECKPOINT
 
-export const removeColumnTypeFromPath = (
-  columnPath: string,
-  type: string,
-  fileSep: string
-) =>
-  columnPath.startsWith(type + fileSep)
+export const removeColumnTypeFromPath = (columnPath: string, type: string) =>
+  columnPath.startsWith(type + FILE_SEPARATOR)
     ? columnPath.slice(type.length + 1)
     : columnPath
 
-export const getFullValuePath = (
-  type: string,
-  columnPath: string,
-  fileSep: string
-) => type + fileSep + columnPath
+export const getFullValuePath = (type: string, columnPath: string) =>
+  type + FILE_SEPARATOR + columnPath
 
 export const cleanupOldOrderValue = (
-  value: { metric: string; param: string } | CustomPlotsOrderValue,
-  fileSep: string
+  value: { metric: string; param: string } | CustomPlotsOrderValue
 ): CustomPlotsOrderValue => ({
   // previous column paths have the "TYPE:" prefix
-  metric: removeColumnTypeFromPath(value.metric, ColumnType.METRICS, fileSep),
-  param: removeColumnTypeFromPath(value.param, ColumnType.PARAMS, fileSep),
+  metric: removeColumnTypeFromPath(value.metric, ColumnType.METRICS),
+  param: removeColumnTypeFromPath(value.param, ColumnType.PARAMS),
   // previous values didn't have a type
   type: (value as CustomPlotsOrderValue).type || CustomPlotType.METRIC_VS_PARAM
 })
+
+export const getCustomPlotIds = (
+  customPlotVals: CustomPlotsOrderValue[],
+  customPlotType: CustomPlotType
+): Set<string> => {
+  const plotIds: Set<string> = new Set()
+
+  for (const { type, metric, param } of customPlotVals) {
+    if (type === customPlotType) {
+      plotIds.add(getCustomPlotId(metric, param))
+    }
+  }
+
+  return plotIds
+}
+
+export const getCustomPlotPathsFromColumns = (
+  columns: Column[]
+): { metrics: string[]; params: string[] } => {
+  const metrics = []
+  const params = []
+
+  for (const { path, type } of columns) {
+    if (type === ColumnType.METRICS) {
+      metrics.push(removeColumnTypeFromPath(path, ColumnType.METRICS))
+    } else if (type === ColumnType.PARAMS) {
+      params.push(removeColumnTypeFromPath(path, ColumnType.PARAMS))
+    }
+  }
+  return { metrics, params }
+}
+
+export const checkForMetricVsParamPlotOptions = (
+  columns: Column[],
+  customPlotOrder: CustomPlotsOrderValue[]
+  // eslint-disable-next-line sonarjs/cognitive-complexity
+): boolean => {
+  const { metrics, params } = getCustomPlotPathsFromColumns(columns)
+  const plotIds = getCustomPlotIds(
+    customPlotOrder,
+    CustomPlotType.METRIC_VS_PARAM
+  )
+
+  for (const metric of metrics) {
+    for (const param of params) {
+      if (!plotIds.has(getCustomPlotId(metric, param))) {
```

Comment:
I tried a couple things but, as far as I can tell, as long as there as an if statement inside these two for loops, I get the complexity warning. Only solutions that come to mind that don't require the if statement require going through the entire loop which I was trying to avoid 🤔 

I ended up adding a cognitive-complexity comment for now but is there a better way to do this?

Reply:
In order to pass the check you need to remove the inner loop. Something like 

```
const hasCombination = (
  plotIds: Set<string>,
  metric: string,
  params: string[]
): boolean => params.some(param => !plotIds.has(getCustomPlotId(metric, param)))

export const checkForMetricVsParamPlotOptions = (
  columns: Column[],
  customPlotOrder: CustomPlotsOrderValue[]
): boolean => {
  const { metrics, params } = getCustomPlotPathsFromColumns(columns)
  const plotIds = getCustomPlotIds(
    customPlotOrder,
    CustomPlotType.METRIC_VS_PARAM
  )

  for (const metric of metrics) {
    if (hasCombination(plotIds, metric, params)) {
      return true
    }
  }

  return false
}
```

However, whether or not this is more readable is debatable.

Maybe we should move complexity to 6 in both sonarjs and code climate. 

WDYT?

cc @sroy3

Reply:
Raised https://github.com/iterative/vscode-dvc/pull/3556 to discuss.

Reply:
One option could be is to create a function that is an iterator and is producing one by one a Cartesian product of metrics and params. It will be used in two places instead of `const { metrics, params } = getCustomPlotPathsFromColumns(columns)` and two `for loops`:

```
for plotId in allMetricVsParamsPlotIds():
   if ...
```